Implement #9650 Add parameter hooks to inventory plugin iocage

vbotka opened the pull request:
##### SUMMARY

Implement #9650 Add parameter hooks to inventory plugin iocage.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME

iocage.py

##### ADDITIONAL INFORMATION

Tested in Ubuntu 24.04

felixfontein commented on plugins/inventory/iocage.py:
```diff
@@ -71,6 +71,15 @@
+              - The variable C(iocage_hooks) is not created if O(hooks) is empty.
+            type: list
+            elements: path
+            version_added: 10.3.0
```
10.3.0 is already out.
```suggestion
            version_added: 10.4.0
```

felixfontein commented on plugins/inventory/iocage.py:
```diff
@@ -286,6 +308,50 @@ def get_inventory(self, path):
+                        p = Popen(cmd_cat_hook, stdout=PIPE, stderr=PIPE, env=my_env)
+                        stdout, stderr = p.communicate()
+                        if p.returncode != 0:
+                            iocage_hooks.append('-')
```
Maybe a warning should be printed in this case? (Or the behavior should be configurable - ignore, warn, error.)

vbotka replied:
The jails may be heterogeneous, and a hook that works for one jail may not work for the other. I want to keep the spirit of silently ignoring `No such file` or any other error:

1) I don't want to complicate the use case where different jails use different hooks or no hooks at all. Just list all hooks and let the *compose* option pick what is needed.

2) The dash "-" is used in *iocage* to represent a missing value. See for example [ioc_list.py#L259](https://github.com/iocage/iocage/blob/master/iocage_lib/ioc_list.py#L259) or [ioc_list.py#L276](https://github.com/iocage/iocage/blob/master/iocage_lib/ioc_list.py#L276). We've already used it too:

```python
if iocage_ip4_dict['ip4']:
    iocage_ip4 = ','.join([d['ip'] for d in iocage_ip4_dict['ip4']])
else:
    iocage_ip4 = '-'
```

3) The admins should be responsible for `intercepting` anything. And they should be used to it, especially in the case of the `hooks`. For example, the [/etc/dhclient-enter-hooks and /etc/dhclient-exit-hooks](https://man.freebsd.org/cgi/man.cgi?dhclient-script(8)) silently ignore any failing lines in the scripts. It is expected, that the admin is responsible for checking what a hook is doing. There are also security implications.

We can add the options (ignore, warn, error) later if needed.

felixfontein commented on plugins/inventory/iocage.py:
```diff
@@ -286,6 +308,50 @@ def get_inventory(self, path):
+                            raise AnsibleError(f'Invalid (non unicode) input returned: {e}') from e
+
+                    except Exception:
+                        iocage_hooks.append('-')
```
Same here.

Also, why not using `None` instead of `-`? `None` can never appear as a real value, so you can distinguish "error happened" from "`-` was actually read".

vbotka replied:
The dash "-" is used in *iocage* to represent a missing value. See for example [ioc_list.py#L259](https://github.com/iocage/iocage/blob/master/iocage_lib/ioc_list.py#L259) or [ioc_list.py#L276](https://github.com/iocage/iocage/blob/master/iocage_lib/ioc_list.py#L276). We've already used it too:

```python
if iocage_ip4_dict['ip4']:
    iocage_ip4 = ','.join([d['ip'] for d in iocage_ip4_dict['ip4']])
else:
    iocage_ip4 = '-'
```

felixfontein reviewed (commented):
Thanks for your contribution!

I'm wondering a bit about the term `hook`. Is this some nomenclature that's used in the `iocage` surroundings for such things? If not, maybe another name would be better. I'm not very good at coming up with names, but maybe something like `read_file_contents`, resulting in `iocage_file_contents` being set?

vbotka commented:
> I'm wondering a bit about the term `hook`. Is this some nomenclature that's used in the iocage surroundings for such things? If not, maybe another name would be better.

The nomenclature is general. A `hook` is a `method to intercept `. For example,

* [Linux kernel hooking library](https://github.com/h33p/kernel-hook)
* [Using ftrace to hook to functions](https://www.kernel.org/doc/html/v4.17/trace/ftrace-uses.html)
* [Package maintainer scripts and hooks](https://kernel-team.pages.debian.net/kernel-handbook/ch-update-hooks.html)

In particular, [dhclient-script](https://man.freebsd.org/cgi/man.cgi?dhclient-script) says:

> If local customizations are needed, they should be possible using the enter and exit hooks ... After  all  processing  has completed, /sbin/dhclient-script checks for the presence of an executable **/etc/dhclient-exit-hooks** script.

There is no better name.

felixfontein commented:
But generally hooks are **running code**. I would expect to be able to provide a script that runs something when reading "hook". So I think the name "hook" is not a good choice in this context.

russoz reviewed (approved):
LGTM as far as I can tell.

One thing I noticed, but probably for a different PR, is that the YAML indentation is not consistent within the file (or with other plugins for that matter).

vbotka commented:
> ... YAML indentation is not consistent within the file ...

Thank you for the review!

* I fixed the 4-space indentation in the DOCUMENTATION block. There is the 2-space indentation in the EXAMPLES block. Both blocks pass the YAML lint. YAML doesn't enforce the indentation size. The syntax of the YAML blocks is correct.

* I assume the 4 spaces in DOCUMENTATION might be needed to merge the fragments, but I'm not sure when I see the inconsistencies for example in [constructed.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/doc_fragments/constructed.py). The rendering is horrible

```bash
shell> ansible-doc -t inventory community.general.iocage
  ...
   keyed_groups  Add hosts to group based on the values of a variable.
        default: []
        elements: dict
        type: list
        suboptions:

           default_value          The default value when the host variable's value is an empty
                          string.
                          This option is mutually exclusive with
                          `keyed_groups[].trailing_separator'.
          default: null
          type: str
          added in: version 2.12 of ansible-core

           key          The key from input dictionary used to generate groups.
          default: null
  ...
```

felixfontein commented:
> I fixed the 4-space indentation in the DOCUMENTATION block. There is the 2-space indentation in the EXAMPLES block. Both blocks pass the YAML lint. YAML doesn't enforce the indentation size. The syntax of the YAML blocks is correct.

The 2-space indentation is a kind of new requirement of community.general that's not yet checked by CI. (Also because ansible-test is trying to make collection maintainer's life difficult by not adding any features that are useful only for collections.)

The indentation in the YAML has no influence on the rendering in the ansible-doc text output. This looks like a bug in ansible-doc which applies spacing the wrong way in suboptions for the first line of the first paragraph. I'm also getting the same broken output in `ansible-doc -t inventory ansible.builtin.constructed`. (There are also other problems there, like line wrapping sometimes being broken for the first line of a paragraph, also for top-level options.)

vbotka commented:
> The 2-space indentation is a kind of new requirement of community.general

I fixed the 2-space indentation of the DOCUMENTATION block. Thank you.

felixfontein commented:
I created a PR to fix the formatting problem in ansible-doc: https://github.com/ansible/ansible/pull/84690

felixfontein commented on changelogs/fragments/9651-iocage-inventory-hooks.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - iocage inventory plugin - the new parameter ``hooks`` of the plugin is a list of files inside a jail that provide configuration parameters for the inventory. The inventory plugin reads the files from the jails and put the contents into the items of created variable ``iocage_hooks`` (https://github.com/ansible-collections/community.general/issues/9650, https://github.com/ansible-collections/co ...
```
I think there's one thing left (sorry I didn't notice earlier), then it's done:
```suggestion
  - iocage inventory plugin - the new parameter ``hooks_results`` of the plugin is a list of files inside a jail that provide configuration parameters for the inventory. The inventory plugin reads the files from the jails and put the contents into the items of created variable ``iocage_hooks`` (https://github.com/ansible-collections/community.general/issues/9650, https://github.com/ansible-collections/community.general/pull/9651).
```

vbotka replied:
My fault, actually.  Thank you.
